packages/vite-plugin/src/index.ts

```diff
@@ -31,37 +32,64 @@ export function vanillaExtractPlugin({
   let useRuntime = false;
   const cssMap = new Map<string, string>();
 
+  let virtualExt: string;
+
   return {
     name: 'vanilla-extract',
     enforce: 'pre',
+    config(_userConfig, env) {
+      useRuntime =
+        devStyleRuntime === 'vanilla-extract' && env.command === 'serve';
+
+      const include = useRuntime ? ['@vanilla-extract/css/injectStyles'] : [];
+
+      return {
+        optimizeDeps: { include },
+        ssr: {
+          external: [
+            '@vanilla-extract/css',
+            '@vanilla-extract/css/fileScope',
+            '@vanilla-extract/css/adapter',
+          ],
+        },
+      };
+    },
     configResolved(resolvedConfig) {
       config = resolvedConfig;
-      useRuntime =
-        devStyleRuntime === 'vanilla-extract' && config.command === 'serve';
+
+      virtualExt = `.vanilla.${useRuntime ? 'js' : 'css'}?hash=`;
 
       packageInfo = getPackageInfo(config.root);
     },
     resolveId(id) {
-      if (virtualCssFileFilter.test(id)) {
-        const { fileName, source } = getSourceFromVirtualCssFile(id);
-
-        // resolveId shouldn't really cause a side-effect however custom module meta isn't currently working
-        // This is a hack work around until https://github.com/vitejs/vite/issues/3240 is resolved
-        const shortHashFileName = normalizePath(
-          `${fileName}?hash=${hash(source)}`,
-        );
-        cssMap.set(shortHashFileName, source);
-
-        return shortHashFileName;
+      if (id.indexOf(virtualExt) > 0) {
+        return id;
       }
     },
     load(id) {
-      if (cssMap.has(id)) {
-        const css = cssMap.get(id);
+      const extensionIndex = id.indexOf(virtualExt);
+
+      if (extensionIndex > 0) {
+        const fileScopeId = id.substring(0, extensionIndex);
+
+        if (cssMap.has(fileScopeId)) {
```

Comment:
Wondering if we should throw an explicit error if this condition is false. It should never really occur yeah?